File: ABMC.RANFE.LIGHT_SQL/ABMC.RANFE.LIGHT_SQL/ReformaTributaria/Leiaute/JsonMinimo.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;

namespace ABMC.RANFE.LIGHT_SQL.ReformaTributaria.Leiaute
{
    /// <summary>
    /// Extração dos layouts do safx-reforma-layout.json sem depender de biblioteca
    /// externa. Lê só o necessário: nome do layout, ordem dos campos, tamanho e o
    /// marcador "ordemConfirmada".
    /// </summary>
    internal sealed class JsonMinimo
    {
        private readonly string _texto;
        private int _pos;

        private JsonMinimo(string texto)
        {
            _texto = texto;
        }

        internal static Dictionary<string, LayoutReforma.Layout> Layouts(string json)
        {
            try
            {
                JsonMinimo parser = new JsonMinimo(json);
                Dictionary<string, object> mapa = (Dictionary<string, object>)parser.Valor();
                Dictionary<string, object> declarados = (Dictionary<string, object>)mapa["layouts"];

                Dictionary<string, LayoutReforma.Layout> resultado =
                    new Dictionary<string, LayoutReforma.Layout>(StringComparer.Ordinal);

                foreach (KeyValuePair<string, object> e in declarados)
                {
                    Dictionary<string, object> corpo = (Dictionary<string, object>)e.Value;
                    List<LayoutReforma.Campo> campos = new List<LayoutReforma.Campo>();
                    foreach (object o in (List<object>)corpo["fields"])
                    {
                        Dictionary<string, object> f = (Dictionary<string, object>)o;
                        campos.Add(new LayoutReforma.Campo(
                            Texto(f, "campo"),
                            Texto(f, "origem"),
                            Texto(f, "tipo"),
                            Texto(f, "tamanho")));
                    }
                    object confirmada = Valor(corpo, "ordemConfirmada");
                    object oficial = Valor(corpo, "fieldCountOficial");
                    resultado[e.Key] = new LayoutReforma.Layout(
                        e.Key,
                        campos,
                        confirmada is bool && (bool)confirmada,
                        oficial is double ? (int)(double)oficial : campos.Count);
                }
                return resultado;
            }
            catch (Exception ex)
            {
                throw new IOException("Falha ao ler " + LayoutReforma.NomeArquivo + ": " + ex.Message, ex);
            }
        }

        private static object Valor(Dictionary<string, object> mapa, string chave)
        {
            object v;
            return mapa.TryGetValue(chave, out v) ? v : null;
        }

        private static string Texto(Dictionary<string, object> mapa, string chave)
        {
            object v = Valor(mapa, chave);
            return v == null ? "" : Convert.ToString(v, CultureInfo.InvariantCulture);
        }

        private object Valor()
        {
            Espacos();
            char c = _texto[_pos];
            switch (c)
            {
                case '{': return Objeto();
                case '[': return Lista();
                case '"': return Cadeia();
                case 't': _pos += 4; return true;
                case 'f': _pos += 5; return false;
                case 'n': _pos += 4; return null;
                default: return Numero();
            }
        }

        private Dictionary<string, object> Objeto()
        {
            Dictionary<string, object> mapa = new Dictionary<string, object>(StringComparer.Ordinal);
            _pos++;
            Espacos();
            if (_texto[_pos] == '}') { _pos++; return mapa; }
            while (true)
            {
                Espacos();
                string chave = Cadeia();
                Espacos();
                _pos++; // :
                mapa[chave] = Valor();
                Espacos();
                if (_texto[_pos++] == '}') return mapa;
            }
        }

        private List<object> Lista()
        {
            List<object> itens = new List<object>();
            _pos++;
            Espacos();
            if (_texto[_pos] == ']') { _pos++; return itens; }
            while (true)
            {
                itens.Add(Valor());
                Espacos();
                if (_texto[_pos++] == ']') return itens;
            }
        }

        private string Cadeia()
        {
            StringBuilder sb = new StringBuilder();
            _pos++;
            while (true)
            {
                char c = _texto[_pos++];
                if (c == '"') return sb.ToString();
                if (c != '\\') { sb.Append(c); continue; }
                char e = _texto[_pos++];
                switch (e)
                {
                    case 'n': sb.Append('\n'); break;
                    case 't': sb.Append('\t'); break;
                    case 'r': sb.Append('\r'); break;
                    case 'u':
                        sb.Append((char)int.Parse(_texto.Substring(_pos, 4),
                            NumberStyles.HexNumber, CultureInfo.InvariantCulture));
                        _pos += 4;
                        break;
                    default: sb.Append(e); break;
                }
            }
        }

        private double Numero()
        {
            int inicio = _pos;
            while (_pos < _texto.Length && "-+.eE0123456789".IndexOf(_texto[_pos]) >= 0) _pos++;
            return double.Parse(_texto.Substring(inicio, _pos - inicio), CultureInfo.InvariantCulture);
        }

        private void Espacos()
        {
            while (_pos < _texto.Length && char.IsWhiteSpace(_texto[_pos])) _pos++;
        }
    }
}

File: código-fonte-reforma/src/main/java/com/caldeira/service/JsonMinimo.java
package com.caldeira.service;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Extração dos layouts do safx-reforma-layout.json sem dependência externa (o
 * pom do projeto não tem nenhuma). Lê só o necessário: nome do layout, ordem
 * dos campos e o marcador "ordemConfirmada".
 */
final class JsonMinimo {

    private final String texto;
    private int pos;

    private JsonMinimo(String texto) {
        this.texto = texto;
    }

    static Map<String, LayoutReforma.Layout> layouts(String json) throws IOException {
        try {
            JsonMinimo parser = new JsonMinimo(json);
            Object raiz = parser.valor();
            Map<String, Object> mapa = comoMapa(raiz);
            Map<String, Object> declarados = comoMapa(mapa.get("layouts"));

            Map<String, LayoutReforma.Layout> resultado = new LinkedHashMap<>();
            for (Map.Entry<String, Object> e : declarados.entrySet()) {
                Map<String, Object> corpo = comoMapa(e.getValue());
                List<LayoutReforma.Campo> campos = new ArrayList<>();
                for (Object o : comoLista(corpo.get("fields"))) {
                    Map<String, Object> f = comoMapa(o);
                    campos.add(new LayoutReforma.Campo(
                            String.valueOf(f.get("campo")),
                            String.valueOf(f.get("origem")),
                            String.valueOf(f.get("tipo")),
                            f.get("tamanho") == null ? "" : String.valueOf(f.get("tamanho"))));
                }
                boolean confirmada = Boolean.TRUE.equals(corpo.get("ordemConfirmada"));
                int oficial = corpo.get("fieldCountOficial") instanceof Double
                        ? ((Double) corpo.get("fieldCountOficial")).intValue()
                        : campos.size();
                resultado.put(e.getKey(),
                        new LayoutReforma.Layout(e.getKey(), campos, confirmada, oficial));
            }
            return resultado;
        } catch (RuntimeException ex) {
            throw new IOException("Falha ao ler " + LayoutReforma.NOME_ARQUIVO + ": " + ex.getMessage(), ex);
        }
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> comoMapa(Object o) {
        return (Map<String, Object>) o;
    }

    @SuppressWarnings("unchecked")
    private static List<Object> comoLista(Object o) {
        return (List<Object>) o;
    }

    private Object valor() {
        espacos();
        char c = texto.charAt(pos);
        switch (c) {
            case '{': return objeto();
            case '[': return lista();
            case '"': return string();
            case 't': pos += 4; return Boolean.TRUE;
            case 'f': pos += 5; return Boolean.FALSE;
            case 'n': pos += 4; return null;
            default: return numero();
        }
    }

    private Map<String, Object> objeto() {
        Map<String, Object> mapa = new LinkedHashMap<>();
        pos++;
        espacos();
        if (texto.charAt(pos) == '}') { pos++; return mapa; }
        while (true) {
            espacos();
            String chave = string();
            espacos();
            pos++; // :
            mapa.put(chave, valor());
            espacos();
            if (texto.charAt(pos++) == '}') return mapa;
        }
    }

    private List<Object> lista() {
        List<Object> itens = new ArrayList<>();
        pos++;
        espacos();
        if (texto.charAt(pos) == ']') { pos++; return itens; }
        while (true) {
            itens.add(valor());
            espacos();
            if (texto.charAt(pos++) == ']') return itens;
        }
    }

    private String string() {
        StringBuilder sb = new StringBuilder();
        pos++;
        while (true) {
            char c = texto.charAt(pos++);
            if (c == '"') return sb.toString();
            if (c != '\\') { sb.append(c); continue; }
            char e = texto.charAt(pos++);
            switch (e) {
                case 'n': sb.append('\n'); break;
                case 't': sb.append('\t'); break;
                case 'r': sb.append('\r'); break;
                case 'u':
                    sb.append((char) Integer.parseInt(texto.substring(pos, pos + 4), 16));
                    pos += 4;
                    break;
                default: sb.append(e);
            }
        }
    }

    private Double numero() {
        int inicio = pos;
        while (pos < texto.length() && "-+.eE0123456789".indexOf(texto.charAt(pos)) >= 0) pos++;
        return Double.valueOf(texto.substring(inicio, pos));
    }

    private void espacos() {
        while (pos < texto.length() && Character.isWhitespace(texto.charAt(pos))) pos++;
    }
}
